README.md

[![Arduino CI](https://github.com/RobTillaart/TCA9548/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/TCA9548/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/TCA9548/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/TCA9548/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/TCA9548/actions/workflows/jsoncheck.yml)
[![GitHub issues](https://img.shields.io/github/issues/RobTillaart/TCA9548.svg)](https://github.com/RobTillaart/TCA9548/issues)

[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/TCA9548/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/TCA9548.svg?maxAge=3600)](https://github.com/RobTillaart/TCA9548/releases)
[![PlatformIO Registry](https://badges.registry.platformio.org/packages/robtillaart/library/TCA9548.svg)](https://registry.platformio.org/libraries/robtillaart/TCA9548)


# TCA9548

Arduino Library for TCA9548 I2C multiplexer and compatibles.


## Description

Library for the TCA9548 and TCA9548a (PCA9548, PCA9548a, PCA9546, PCA9545, PCA9543) I2C multiplexer.

The library allows you to enable 0 to 7 I2C channels (ports) uniquely or simultaneously.
In fact the TCA9548 is therefore a **switch**, although often named a multiplexer.
A multiplexer is especially useful if you have multiple identical devices that have a fixed address,
a too small address range or if there are address conflicts between different I2C devices.

**Warning**
The library is not tested extensively.

The library caches the channels enabled, and if a channel is enabled,
it will not be enabled again (low level) to optimize performance.

#### I2C 

I2C address of the device itself is 0x70 .. 0x77.
This address can not be used on any of the I2C channels of course.

Note if your first multiplexer is 0x70, you may have an array of 0x71 multiplexers behind it.
(giving you 8 x 8 = 64 I2C buses, a lot of admin overhead and probably performance penalties).

Pull-up resistors are REQUIRED on all upstream and downstream channels.


#### 0.2.0 Breaking change

Version 0.2.0 introduced a breaking change.
You cannot set the pins in **begin()** any more.
This reduces the dependency of processor dependent Wire implementations.
The user has to call **Wire.begin()** and can optionally set the Wire pins 
before calling **begin()**.


#### Compatible devices

This library is expected to work for the following devices: (since 0.2.1)

|  Device    |  Tested  |  Notes  |
|:-----------|:--------:|:-------:|
|  TCA9548s  |    n     |
|  PCA9548   |    n     |  see links below  |
|  PCA9548a  |    n     |
|  PCA9546   |    n     |  see links below  |
|  PCA9545   |    n     |  see links below  |
|  PCA9543   |    n     |  see links below  |

Note: not tested with hardware yet.

There are however small differences, check the data sheets to see the details.
- [difference TCA PCA](https://e2e.ti.com/support/interface-group/interface/f/interface-forum/815758/faq-what-is-the-difference-between-an-i2c-device-with-the-family-name-pca-and-tca)
- https://electronics.stackexchange.com/questions/209616/is-nxps-pca9548a-compatible-with-tis-tca9548a
- https://www.nxp.com/docs/en/application-note/AN262.pdf


#### Related

- https://github.com/RobTillaart/HC4051  (1x8 mux)
- https://github.com/RobTillaart/HC4052  (2x4 mux)
- https://github.com/RobTillaart/HC4053  (3x2 mux)
- https://github.com/RobTillaart/HC4067  (1x16 mux)


## Interface

```cpp
#include "TCA9548.h"
```

#### Constructor

- **TCA9548(const uint8_t deviceAddress, TwoWire \*wire = &Wire)** Constructor.
deviceAddress = 0x70 .. 0x77, wire = Wire or WireN.
- **bool begin(uint8_t dataPin, uint8_t clockPin, uint8_t mask = 0x00)**  Set I2C pins for ESP32.
Set mask of channels to be enabled, default all disabled.
- **bool begin(uint8_t mask = 0x00)**  set mask of channels to be enabled, default all disabled.
- **bool isConnected()** returns true if address of the multiplexer is found on I2C bus.


The derived classes PCA9548/PCA9546 have the same interface, except constructor.
(see #15)

- **PCA9548(const uint8_t deviceAddress, TwoWire \*wire = &Wire)** Constructor.
deviceAddress = 0x70 .. 0x77, wire = Wire or WireN.
- **PCA9546(const uint8_t deviceAddress, TwoWire \*wire = &Wire)** Constructor.
deviceAddress = 0x70 .. 0x77, wire = Wire or WireN.
- **PCA9545(const uint8_t deviceAddress, TwoWire \*wire = &Wire)** Constructor.
deviceAddress = 0x70 .. 0x77, wire = Wire or WireN.
- **PCA9543(const uint8_t deviceAddress, TwoWire \*wire = &Wire)** Constructor.
deviceAddress = 0x70 .. 0x77, wire = Wire or WireN.


#### Find device

- **bool isConnected(uint8_t address)** returns true if arbitrary address is found on I2C bus.
This can be used to verify a certain device is available (or not) on an enabled channel.


#### Channel functions

All "channel functions" return true on success.

- **bool enableChannel(uint8_t channel)** enables channel 0 .. 7 non-exclusive.
Multiple channels can be enabled in parallel.
- **bool disableChannel(uint8_t channel)** disables channel 0 .. 7.
Will not disable other channels.
- **bool selectChannel(uint8_t channel)** enables a single channel 0 .. 7 exclusive.
All other channels will be disabled in the same call, so not before or after.
- **bool isEnabled(uint8_t channel)** returns true if the channel is enabled.
- **bool disableAllChannels()** fast way to disable all.
- **bool setChannelMask(uint8_t mask)** enables 0 or more channels simultaneously with a bit mask.
- **uint8_t getChannelMask()** reads back the bit mask of the channels enabled.


#### Reset

- **void setResetPin(uint8_t resetPin)** sets the pin to reset the chip. (Not tested)
- **void reset()** trigger the reset pin.
- **int getError()** returns the last I2C error.


#### Forced IO

When forced IO is set all writes and read - **getChannelMask()** - will go to the device.
If the flag is set to false it will cache the value of the channels enabled.
This will result in more responsive / faster calls.
Note that writes are only optimized if the channels are already set.

- **void setForced(bool forced = false)** set forced write, slower but more robust.
  - forced == false == fast mode (default).
  - forced == true == slower, robust mode.
- **bool getForced()** returns set flag.


#### Interrupts

The PCA9545 and PCA9543 do support interrupts. 
These two derived classes have implemented

- **uint8_t getInterruptMask()** function that returns a bit mask of interrupts set.


## Error Codes

Not implemented yet, preparation for future.

|  name                   |  value  |  description            |
|:------------------------|:-------:|:------------------------|
|  TCA9548_OK             |   00    |  no error               |
|  TCA9548_ERROR_I2C      |  -10    |  detected an I2C error  |
|  TCA9548_ERROR_CHANNEL  |  -20    |  channel out of range   |


## Future


#### Must

- improve documentation.
- improve error handling.

#### Should

- add examples.
- test test and test.
- write unit test.


#### Could

- set an "always enabled" mask.
  - investigate the consequences!

#### Wont


## Support

If you appreciate my libraries, you can support the development and maintenance.
Improve the quality of the libraries by providing issues and Pull Requests, or
donate through PayPal or GitHub sponsors.

Thank you,





CHANGELOG.md
# Change Log TCA9548

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.2.1] - 2023-12-09
- add derived classes
  - PCA9543 (2 channel), PCA9545 (4 channel), PCA9546 (4 channel)
- add **uint8_t getInterruptMask()**
- fix begin() : remove wire-> begin() as dependency should be outside lib


## [0.2.0] - 2023-12-09
- refactor API, begin()
- update readme.md
- add derived class PCA9548

----

## [0.1.6] - 2023-11-22
- update readme.md
- update keywords.txt

## [0.1.5] - 2023-02-12
- fix #8 ambiguity
- update keywords.txt
- add example **tca9548_search_device.ino**
- update readme.md
- prepare error handling (constants)
- some minor edits.

## [0.1.4] - 2023-01-23
- update GitHub actions
- update license 2023
- update readme.md
- update keywords.txt
- add **void disableAllChannels()**
- all channel functions return bool, true on success.
- minor edits

## [0.1.3] - 2022-11-26
- Add RP2040 support to build-CI.
- Add CHANGELOG.md
- update readme.md
- move code to .cpp

## [0.1.2] - 2021-12-28
- update license
- minor edits

## [0.1.1] - 2021-11-19
- fix reset code (from datasheet)
- implemented forced IO
- update build-CI, readme.md, badges

## [0.1.0] - 2021-03-16
- initial version



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.0
//    DATE: 2021-03-16
// PURPOSE: unit tests for the TCA9548 I2C multiplexer
//          https://github.com/RobTillaart/TCA9548
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)

/*
  most unit tests will test for fail 
  as there is no sensor connected
  and there is no mockup.
  
  It appears that Wire.write does not fail without sensor...
*/

#include <ArduinoUnitTests.h>

#include "Arduino.h"
#include "TCA9548.h"

int expect;  // TODO needed as there seems a problem with 8 bit comparisons (char?)

uint32_t start, stop;

unittest_setup()
{
}

unittest_teardown()
{
}

unittest(test_begin)
{
  TCA9548 tca(0x70);

  bool b = tca.begin();
  assertEqual(b, true);

  assertTrue(tca.isConnected());

}

unittest(test_enable)
{
  TCA9548 tca(0x70);

  bool b = tca.begin();
  assertEqual(b, true);

  uint8_t mask = 0x00;
  for (int chan = 0; chan < 8; chan++)
  {
    tca.enableChannel(chan);
    assertTrue(tca.isEnabled());
  }
  assertEqual(0xFF, tca.getChannelMask());
  tca.setChannelMask(0x00);
  assertEqual(0x00, tca.getChannelMask());
}


unittest(test_select)
{
  TCA9548 tca(0x70);

  bool b = tca.begin();
  assertEqual(b, true);

  uint8_t mask = 0x00;
  for (int chan = 0; chan < 8; chan++)
  {
    tca.selectChannel(chan);
    assertTrue(tca.isEnabled());
  }
  assertEqual(0x80, tca.getChannelMask());
  tca.setChannelMask(0x00);
  assertEqual(0x00, tca.getChannelMask());
}


unittest_main()

// --------

TCA9548.cpp
//
//    FILE: TCA9548.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.2.1
//    DATE: 2021-03-16
// PURPOSE: Arduino Library for TCA9548 I2C multiplexer and compatibles.


#include "TCA9548.h"


TCA9548::TCA9548(uint8_t deviceAddress, TwoWire *wire)
{
  _address  = deviceAddress;
  _wire     = wire;
  _mask     = 0x00;
  _resetPin = -1;
  _forced   = false;
  _error    = TCA9548_OK;
  _channels = 8;
}


bool TCA9548::begin(uint8_t mask)
{
  if (! isConnected()) return false;
  setChannelMask(mask);
  return true;
}


bool TCA9548::isConnected()
{
  _wire->beginTransmission(_address);
  return ( _wire->endTransmission() == 0);
}


bool TCA9548::isConnected(uint8_t address)
{
  _wire->beginTransmission(address);
  return ( _wire->endTransmission() == 0);
}


uint8_t TCA9548::channelCount()
{
  return _channels;
}


bool TCA9548::enableChannel(uint8_t channel)
{
  if (channel >= _channels) return false;
  if (!isEnabled(channel))
  {
    setChannelMask(_mask | (0x01 << channel));
  }
  return true;
}


bool TCA9548::disableChannel(uint8_t channel)
{
  if (channel >= _channels) return false;
  if (!isEnabled(channel))
  {
    setChannelMask(_mask & ~(0x01 << channel));
  }
  return true;
}


bool TCA9548::selectChannel(uint8_t channel)
{
  if (channel >= _channels) return false;
  setChannelMask(0x01 << channel);
  return true;
}


bool TCA9548::isEnabled(uint8_t channel)
{
  if (channel >= _channels) return false;
  return (_mask & (0x01 << channel));
}


bool TCA9548::disableAllChannels()
{
  return setChannelMask(0x00);
}


bool TCA9548::setChannelMask(uint8_t mask)
{
  if ((_mask == mask) && (not _forced)) return true;
  _mask = mask;
  _wire->beginTransmission(_address);
  _wire->write(_mask);
  _error = _wire->endTransmission();
  return (_error == 0);
}


uint8_t TCA9548::getChannelMask()
{
  if (_forced)  //  read from device.
  {
    _wire->requestFrom(_address, (uint8_t)1);
    _mask = _wire->read();
  }
  return _mask;
}


void TCA9548::setResetPin(uint8_t resetPin)
{
  _resetPin = resetPin;
  pinMode(_resetPin, OUTPUT);
  digitalWrite(_resetPin, HIGH);  //  page 3 HIGH is normal operation
}


void TCA9548::reset()
{
  digitalWrite(_resetPin, LOW);
  delayMicroseconds(1);           //  datasheet page 6 & 7 - 500 ns
  digitalWrite(_resetPin, HIGH);
}


void TCA9548::setForced(bool forced)
{
  _forced = forced;
};


bool TCA9548::getForced()
{
  return _forced;
};


int TCA9548::getError()
{
  int error = _error;
  _error = TCA9548_OK;
  return error;
}


/////////////////////////////////////////////////////////////
//
//  PCA9548
//
PCA9548::PCA9548(uint8_t deviceAddress, TwoWire *wire) : TCA9548(deviceAddress, wire)
{
  _channels = 8;
}


/////////////////////////////////////////////////////////////
//
//  PCA9546
//
PCA9546::PCA9546(uint8_t deviceAddress, TwoWire *wire) : TCA9548(deviceAddress, wire)
{
  _channels = 4;
}

uint8_t PCA9546::getChannelMask()
{
  if (_forced)  //  read from device.
  {
    _wire->requestFrom(_address, (uint8_t)1);
    _mask = _wire->read();
  }
  return _mask &= 0x0F;
}


/////////////////////////////////////////////////////////////
//
//  PCA9545
//
PCA9545::PCA9545(uint8_t deviceAddress, TwoWire *wire) : TCA9548(deviceAddress, wire)
{
  _channels = 4;
}

uint8_t PCA9545::getChannelMask()
{
  if (_forced)  //  read from device.
  {
    _wire->requestFrom(_address, (uint8_t)1);
    _mask = _wire->read();
  }
  return _mask &= 0x0F;
}

uint8_t PCA9545::getInterruptMask()
{
  _wire->requestFrom(_address, (uint8_t)1);
  uint8_t mask = _wire->read();
  mask >>= 4;
  return mask;
}


/////////////////////////////////////////////////////////////
//
//  PCA9548
//
PCA9543::PCA9543(uint8_t deviceAddress, TwoWire *wire) : TCA9548(deviceAddress, wire)
{
  _channels = 2;
}

uint8_t PCA9543::getChannelMask()
{
  if (_forced)  //  read from device.
  {
    _wire->requestFrom(_address, (uint8_t)1);
    _mask = _wire->read();
  }
  return _mask &= 0x03;
}

uint8_t PCA9543::getInterruptMask()
{
  _wire->requestFrom(_address, (uint8_t)1);
  uint8_t mask = _wire->read();
  mask >>= 4;
  return mask;
}


//  -- END OF FILE --



TCA9548.h
#pragma once
//
//    FILE: TCA9548.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.2.1
//    DATE: 2021-03-16
// PURPOSE: Arduino Library for TCA9548 I2C multiplexer and compatibles.
//     URL: https://github.com/RobTillaart/TCA9548


#include "Arduino.h"
#include "Wire.h"


#define TCA9548_LIB_VERSION             (F("0.2.1"))


//  ERROR CODES (to be elaborated)
#define TCA9548_OK                      0
#define TCA9548_ERROR_I2C               -10  //  not used yet
#define TCA9548_ERROR_CHANNEL           -20  //  not used yet



class TCA9548
{
public:
  //  deviceAddress = 0x70 .. 0x77
  TCA9548(uint8_t deviceAddress, TwoWire *wire = &Wire);

  bool    begin(uint8_t mask = 0x00);         //  default no channels enabled
  bool    isConnected();                      //  find multiplexer on I2C bus
  bool    isConnected(uint8_t address);       //  find any address on I2C bus


  //  channel = 0..channelCount()-1
  uint8_t channelCount();
  bool    enableChannel(uint8_t channel);     //  enable this channel non exclusive
  bool    disableChannel(uint8_t channel);
  bool    selectChannel(uint8_t channel);     //  enable only this channel
  bool    isEnabled(uint8_t channel);
  bool    disableAllChannels();


  //  mask = 0x00 .. 0xFF - every bit is a channel.
  //         although not for derived types.
  //  note 1:  these are set simultaneously.
  //  note 2:  these may include interrupt bits (in derived classes)
  bool    setChannelMask(uint8_t mask);
  uint8_t getChannelMask();

  //  reset
  void    setResetPin(uint8_t resetPin);
  void    reset();     //  trigger reset pin

  //  set forced IO (default false)
  void    setForced(bool forced = false);
  bool    getForced();

  int     getError();


protected:
  uint8_t   _address;
  TwoWire*  _wire;
  uint8_t   _mask;             //  caching mask = status of channels
  uint8_t   _resetPin;         //  default not set == -1 (255)
  bool      _forced;
  int       _error;
  uint8_t   _channels;         //  PCA954x support.
  uint8_t   _type;             //  PCA954x support.
};


/////////////////////////////////////////////////////////////
//
//  DERIVED CLASS
//
//  see - https://github.com/RobTillaart/TCA9548/issues/15
//

//  SWITCH, 8 channel + reset
class PCA9548 : public TCA9548
{
public:
  PCA9548(uint8_t deviceAddress, TwoWire *wire = &Wire);
};


//  SWITCH, 4 channel + reset
class PCA9546 : public TCA9548
{
public:
  PCA9546(uint8_t deviceAddress, TwoWire *wire = &Wire);
  uint8_t getChannelMask();
};


//  DEVICES WITH INTERRUPT 

//  SWITCH, 4 channel + reset + interrupt
class PCA9545 : public TCA9548
{
public:
  PCA9545(uint8_t deviceAddress, TwoWire *wire = &Wire);
  uint8_t getChannelMask();
  uint8_t getInterruptMask();
};


//  SWITCH, 2 channel + reset + interrupt
class PCA9543 : public TCA9548
{
public:
  PCA9543(uint8_t deviceAddress, TwoWire *wire = &Wire);
  uint8_t getChannelMask();
  uint8_t getInterruptMask();
};


//  -- END OF FILE --

